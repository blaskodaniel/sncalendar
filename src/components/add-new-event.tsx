import React, { useContext } from 'react'
import { createStyles, Fab, makeStyles, Theme } from '@material-ui/core'
import AddIcon from '@material-ui/icons/Add'
import { SharedContext } from '../context/shared-context'

/**
 * Style for component with Material UI
 */
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    fabAdd: {
      position: 'fixed',
      bottom: theme.spacing(2),
      right: theme.spacing(2),
    },
    extendedIcon: {
      marginRight: theme.spacing(1),
    },
  }),
)

const AddNewEvent: React.FunctionComponent = () => {
  const classes = useStyles()
  const sharedcontext = useContext(SharedContext)

  return (
    <Fab
      color="secondary"
      variant="extended"
      onClick={() => {
        // New event
        sharedcontext.setOpennewmodal(true)
      }}
      aria-label="Add new"
      className={classes.fabAdd}>
      <AddIcon className={classes.extendedIcon} />
      Add new
    </Fab>
  )
}

export default AddNewEvent
