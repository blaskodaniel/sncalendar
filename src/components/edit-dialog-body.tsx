import React, { useContext } from 'react'
import { DialogContent, DialogTitle } from '@material-ui/core'
import { EditView } from '@sensenet/controls-react'
import { DialogProps } from '@material-ui/core/Dialog'
import { ConstantContent } from '@sensenet/client-core'
import { ContentType } from '@sensenet/default-content-types'
import { useRepository } from '../hooks/use-repository'
import { useSelectionService } from '../hooks/use-selection-service'
import { CurrentContentContext, CurrentContentProvider } from '../context/current-context'
import CalendarEvent from '../CalendarEvent-type'
import { SharedContext } from '../context/shared-context'

const EditPropertiesDialogBody: React.FunctionComponent<{
  contentId: number
  dialogProps: DialogProps
}> = props => {
  const selectionService = useSelectionService()
  const repo = useRepository()
  const sharedcontext = useContext(SharedContext)

  const onSubmit = async (id: number, content: CalendarEvent) => {
    try {
      const response = await repo.patch({
        idOrPath: id,
        content,
        oDataOptions: {
          select: 'all' as any,
        },
      })
      sharedcontext.setEvent(response.d) // Refresh the view dialog window
      sharedcontext.setRefreshcalendar(!sharedcontext.refreshcalendar) // Refresh the main calendar list
      props.dialogProps.onClose && props.dialogProps.onClose(null as any, 'backdropClick')
    } catch (error) {
      console.log('Error: ', error)
    }
  }

  return (
    <CurrentContentProvider
      idOrPath={props.contentId}
      onContentLoaded={(c: CalendarEvent) => selectionService.activeContent.setValue(c)}
      oDataOptions={{ select: 'all' }}>
      <CurrentContentContext.Consumer>
        {(content: CalendarEvent) =>
          content.Id !== ConstantContent.PORTAL_ROOT.Id && (
            <>
              <DialogTitle>Edit</DialogTitle>
              <DialogContent>
                <EditView
                  content={content as ContentType}
                  repository={repo}
                  handleCancel={() =>
                    props.dialogProps.onClose && props.dialogProps.onClose(null as any, 'backdropClick')
                  }
                  onSubmit={onSubmit}
                />
              </DialogContent>
            </>
          )
        }
      </CurrentContentContext.Consumer>
    </CurrentContentProvider>
  )
}

export default EditPropertiesDialogBody
