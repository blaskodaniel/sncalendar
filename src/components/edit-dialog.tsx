import Dialog, { DialogProps } from '@material-ui/core/Dialog'
import React from 'react'
import CalendarEvent from '../CalendarEvent-type'
import EditPropertiesDialogBody from './edit-dialog-body'

export const EditPropertiesDialog: React.FunctionComponent<{
  dialogProps: DialogProps
  content: CalendarEvent
}> = props => {
  return (
    <Dialog open={props.dialogProps.open} fullScreen aria-labelledby="event-dialog">
      <EditPropertiesDialogBody contentId={props.content.Id} dialogProps={props.dialogProps} />
    </Dialog>
  )
}
