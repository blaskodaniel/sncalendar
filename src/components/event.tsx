import React, { useContext } from 'react'
import moment from 'moment'
import { Avatar, createStyles, List, ListItem, ListItemAvatar, ListItemText, makeStyles } from '@material-ui/core'
import { User } from '@sensenet/default-content-types'
import CalendarEvent from '../CalendarEvent-type'
import defavatar from '../assets/avatar-default.png'
import { useRepository } from '../hooks/use-repository'
import { SharedContext } from '../context/shared-context'

const useStyles = makeStyles(() =>
  createStyles({
    parentlistelement: {
      display: 'flex',
      flexDirection: 'row-reverse',
      minWidth: '350px',
      border: '1px solid rgba(0, 0, 0, 0.12)',
      borderRadius: '2%',
      marginBottom: '7px',
      '&:hover': {
        backgroundColor: '#bad892',
        cursor: 'pointer',
      },
    },
    alldayevent: {
      backgroundColor: '#d3daff',
    },
    simpleevent: {
      backgroundColor: 'azure',
    },
  }),
)

export interface EventComponentProps {
  event: CalendarEvent[]
}

const EventComponent: React.FunctionComponent<EventComponentProps> = props => {
  const classes = useStyles()
  const repo = useRepository()
  const sharedcontext = useContext(SharedContext)
  const timeAndLocationpart = (event: CalendarEvent) => {
    const start = moment(new Date(event.StartDate as string)).format('HH:mm')
    const end = moment(new Date(event.EndDate as string)).format('HH:mm')
    const isAllday = event.AllDay
    const location = event.Location
    const timeDisplay = isAllday ? 'all day' : ` ${start}-${end}`
    return `${timeDisplay} at ${location}`
  }

  return (
    <List>
      {props.event.map(event => {
        return (
          <div key={event.Id}>
            <ListItem
              onClick={() => {
                sharedcontext.setEvent(event)
                sharedcontext.setOpendisplaymodal(true)
              }}
              className={`${classes.parentlistelement} ${event.AllDay ? classes.alldayevent : classes.simpleevent}`}>
              <ListItemAvatar>
                <Avatar
                  alt="Avatar"
                  src={
                    (event.CreatedBy as User).Avatar!.Url === ''
                      ? defavatar
                      : repo.configuration.repositoryUrl + (event.CreatedBy as User).Avatar!.Url
                  }
                />
              </ListItemAvatar>
              <ListItemText primary={event.DisplayName} secondary={timeAndLocationpart(event)} />
            </ListItem>
          </div>
        )
      })}
    </List>
  )
}

export default EventComponent
