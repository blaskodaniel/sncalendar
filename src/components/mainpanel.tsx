import React, { useEffect, useState } from 'react'
import { ODataCollectionResponse } from '@sensenet/client-core'
import moment from 'moment'
import sortby from 'lodash.sortby'
import { createStyles, makeStyles } from '@material-ui/styles'
import { List, ListItem, ListItemAvatar } from '@material-ui/core'
import { v1 } from 'uuid'
import { Query } from '@sensenet/query'
import CalendarEvent from '../CalendarEvent-type'
import { useRepository } from '../hooks/use-repository'
import EventComponent from './event'

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      width: '100%',
    },
    dayAvatar: {
      alignSelf: 'flex-start',
      paddingTop: '8px',
    },
    dayname: {
      display: 'block',
      textAlign: 'center',
    },
    daynumber: {
      display: 'block',
      textAlign: 'center',
      fontSize: '20px',
      fontWeight: 400,
    },
    nopadding: {
      padding: '0',
    },
  }),
)

export interface GroupdByAny {
  date: any
  event: CalendarEvent[]
  id: string
}

/**
 * Main component
 */
const MainPanel: React.FunctionComponent = () => {
  const classes = useStyles()
  const repo = useRepository()
  const [data, setData] = useState<GroupdByAny[]>([])

  const groupByDay = function(xs: CalendarEvent[], key: keyof Pick<CalendarEvent, 'StartDate'>) {
    const resultArray: GroupdByAny[] = []
    xs.forEach(event => {
      const findevent = resultArray.find(
        c =>
          moment(new Date(c.date)).format('YYYY-MM-DD') ===
          moment(new Date(event[key] !== undefined ? (event[key] as string) : '')).format('YYYY-MM-DD'),
      )
      if (findevent) {
        findevent.event.push(event)
        const valami = sortby(findevent.event, 'AllDay', ['asc'])
        console.log('sort: ', valami)
      } else {
        if (event[key]) {
          resultArray.push({
            id: v1(),
            date: event[key],
            event: [event],
          })
        }
      }
    })
    return resultArray
  }

  useEffect(() => {
    const loadCalendar = async () => {
      const result: ODataCollectionResponse<CalendarEvent> = await repo.loadCollection({
        path: `/Root/Content/IT/Calendar`,
        oDataOptions: {
          select: [
            'DisplayName',
            'Description',
            'CreationDate',
            'CreatedBy',
            'ModifiedBy',
            'ModificationDate',
            'Icon',
            'Type',
            'Id',
            'Path',
            'Name',
            'Size',
            'Location',
            'StartDate',
            'EndDate',
            'Lead',
            'AllDay',
            'EventUrl',
            'OwnerEmail',
          ] as any,
          query: new Query(q =>
            q.greatherThan('StartDate', '2019-01-01').and.lessThan('StartDate', '2019.12.31'),
          ).toString(),
          orderby: [['StartDate', 'asc']],
          expand: ['CreatedBy', 'ModifiedBy'],
        },
      })

      const groupedby = groupByDay(result.d.results, 'StartDate')
      console.log(groupedby)
      setData(groupedby)
    }

    // Load calendar datas from Repository
    loadCalendar()
  }, [repo])

  return (
    <>
      {data.map(element => {
        return (
          <List key={element.id} className={classes.root}>
            <ListItem className={classes.nopadding}>
              <ListItemAvatar className={classes.dayAvatar}>
                <div>
                  <span className={classes.dayname}>{moment(new Date(element.date)).format('ddd')}</span>
                  <span className={classes.daynumber}>{moment(new Date(element.date)).format('D')}</span>
                </div>
              </ListItemAvatar>
              <EventComponent event={element.event} />
            </ListItem>
          </List>
        )
      })}
    </>
  )
}

export default MainPanel
