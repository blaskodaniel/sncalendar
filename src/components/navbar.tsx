import React from 'react'
import { AppBar, Button, Toolbar, Tooltip, Typography } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { useRepository } from '../hooks/use-repository'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      flexGrow: 1,
    },
    menuButton: {
      marginRight: theme.spacing(2),
    },
    title: {
      flexGrow: 1,
    },
  }),
)
/**
 * Navbar component
 */
export const NavBarComponent: React.FunctionComponent<{ month: string }> = props => {
  const repo = useRepository()
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <AppBar>
        <Toolbar>
          <Typography variant="h6" className={classes.title}>
            {props.month}
          </Typography>
          <Tooltip title="Return to the Login screen and select another repository">
            <Button color="inherit" onClick={() => repo.authentication.logout()}>
              Log out
            </Button>
          </Tooltip>
        </Toolbar>
      </AppBar>
    </div>
  )
}
