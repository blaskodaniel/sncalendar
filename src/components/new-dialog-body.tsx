import React, { useContext } from 'react'
import { DialogContent, DialogTitle } from '@material-ui/core'
import { NewView } from '@sensenet/controls-react'
import { DialogProps } from '@material-ui/core/Dialog'
import { useRepository } from '../hooks/use-repository'
import CalendarEvent from '../CalendarEvent-type'
import { SharedContext } from '../context/shared-context'

const NewDialogBody: React.FunctionComponent<{
  parentpath: string
  dialogProps: DialogProps
}> = props => {
  const repo = useRepository()
  const sharedcontext = useContext(SharedContext)
  const handleClose = () => props.dialogProps.onClose && props.dialogProps.onClose(null as any, 'backdropClick')

  return (
    <>
      <DialogTitle>Add new event</DialogTitle>
      <DialogContent>
        <NewView
          handleCancel={handleClose}
          repository={repo}
          contentTypeName={CalendarEvent.name}
          path={props.parentpath}
          onSubmit={async (parentPath, content) => {
            try {
              await repo.post({
                contentType: CalendarEvent.name,
                parentPath,
                content,
              })
              sharedcontext.setRefreshcalendar(!sharedcontext.refreshcalendar)
              handleClose()
            } catch (error) {
              console.log(error.message)
            }
          }}
        />
      </DialogContent>
    </>
  )
}

export default NewDialogBody
