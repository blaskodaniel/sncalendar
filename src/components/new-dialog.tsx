import Dialog, { DialogProps } from '@material-ui/core/Dialog'
import React from 'react'
import NewDialogBody from './new-dialog-body'

export const NewDialog: React.FunctionComponent<{
  dialogProps: DialogProps
  parentpath: string
}> = props => {
  return (
    <Dialog open={props.dialogProps.open} fullScreen aria-labelledby="event-dialog">
      <NewDialogBody parentpath={props.parentpath} dialogProps={props.dialogProps} />
    </Dialog>
  )
}
