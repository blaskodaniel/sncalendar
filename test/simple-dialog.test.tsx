import { shallow, ShallowWrapper } from 'enzyme'
import React from 'react'
import Button from '@material-ui/core/Button'
import { DialogComponent, DialogProps } from '../src/components/simple-dialog'

describe('DialogComponent', () => {
  const testprops: DialogProps = {
    open: true,
    title: 'Test title',
    onClose: jest.fn(),
  }
  let wrapper: ShallowWrapper

  beforeEach(() => {
    wrapper = shallow(<DialogComponent {...testprops} />)
  })

  it('DialogComponent snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Open dialog', () => {
    wrapper
      .find(Button)
      .first()
      .simulate('click')
    expect(testprops.onClose).toBeCalledWith(true)
  })

  it('Close dialog', () => {
    wrapper
      .find(Button)
      .last()
      .simulate('click')
    expect(testprops.onClose).toBeCalledWith(false)
  })
})
