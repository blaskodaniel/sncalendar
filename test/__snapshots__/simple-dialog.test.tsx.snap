// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DialogComponent DialogComponent snapshot 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="simple-dialog-title"
  maxWidth="sm"
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="simple-dialog-title"
  >
    Are you sure you want to delete it?
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))>
    <WithStyles(ForwardRef(DialogContentText))
      id="alert-dialog-description"
    >
      Test title
    </WithStyles(ForwardRef(DialogContentText))>
  </WithStyles(ForwardRef(DialogContent))>
  <WithStyles(ForwardRef(DialogActions))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-buttonDefault-1"
      color="secondary"
      onClick={[Function]}
    >
      Yes
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-buttonDefault-1"
      color="default"
      onClick={[Function]}
    >
      No
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogActions))>
</WithStyles(ForwardRef(Dialog))>
`;
